core/state/statedb_hooked.go

```diff
@@ -198,22 +198,46 @@ func (s *hookedStateDB) SetState(address common.Address, key common.Hash, value
 }
 
 func (s *hookedStateDB) SelfDestruct(address common.Address) uint256.Int {
+	var prevCode []byte
+	var prevCodeHash common.Hash
+
+	if s.hooks.OnCodeChange != nil {
+		prevCode = s.inner.GetCode(address)
+		prevCodeHash = s.inner.GetCodeHash(address)
+	}
+
 	prev := s.inner.SelfDestruct(address)
-	if !prev.IsZero() {
-		if s.hooks.OnBalanceChange != nil {
-			s.hooks.OnBalanceChange(address, prev.ToBig(), new(big.Int), tracing.BalanceDecreaseSelfdestruct)
-		}
+
+	if !prev.IsZero() && s.hooks.OnBalanceChange != nil {
+		s.hooks.OnBalanceChange(address, prev.ToBig(), new(big.Int), tracing.BalanceDecreaseSelfdestruct)
+	}
+
+	if len(prevCode) > 0 && s.hooks.OnCodeChange != nil {
+		s.hooks.OnCodeChange(address, prevCodeHash, prevCode, types.EmptyCodeHash, nil)
 	}
+
 	return prev
 }
 
 func (s *hookedStateDB) SelfDestruct6780(address common.Address) (uint256.Int, bool) {
+	var prevCode []byte
+	var prevCodeHash common.Hash
+
+	if s.hooks.OnCodeChange != nil {
+		prevCodeHash = s.inner.GetCodeHash(address)
+		prevCode = s.inner.GetCode(address)
+	}
+
 	prev, changed := s.inner.SelfDestruct6780(address)
-	if !prev.IsZero() && changed {
-		if s.hooks.OnBalanceChange != nil {
-			s.hooks.OnBalanceChange(address, prev.ToBig(), new(big.Int), tracing.BalanceDecreaseSelfdestruct)
-		}
+
+	if !prev.IsZero() && changed && s.hooks.OnBalanceChange != nil {
+		s.hooks.OnBalanceChange(address, prev.ToBig(), new(big.Int), tracing.BalanceDecreaseSelfdestruct)
+	}
+
+	if len(prevCode) > 0 && changed && s.hooks.OnCodeChange != nil {
```

Comment:
Ideally you want these checks to check the presence of the hook first and evaluate other conditions later, since if the hook is disabled (which it will be most of the time), you don't need to check the rest of the conditions. Might not be relevant in most codes, but in hot loops, it could be important.

Reply:
You are right, should be fixed now!